File: scripts/compare_cl1_excel_db.py
#!/usr/bin/env python3
"""对照 CL1 Excel 与 Turso strategy_positions（与 app.services 相同解析规则）。"""
from __future__ import annotations

import os
import sys
from collections import defaultdict
from datetime import date
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from dotenv import load_dotenv

load_dotenv(ROOT / ".env")

EXCEL_DEFAULT = Path(r"D:\展示策略\展示策略\展示策略-沪深300指数增强.xlsx")
SID = "CL1"


def excel_period_counts(file_path: Path) -> tuple[dict[date, int], int]:
    from app.services import _iter_strategy_holdings_excel_batches

    counts: dict[date, int] = defaultdict(int)
    total = 0
    for chunk in _iter_strategy_holdings_excel_batches(str(file_path)):
        for rb, _code, _hw, _iw in chunk:
            counts[rb] += 1
            total += 1
        chunk.clear()
    return dict(counts), total


def db_period_counts() -> tuple[dict[date, int], int, date | None, date | None]:
    from sqlalchemy import text
    from app.db import SessionLocalFactory, init_database
    from app.services import _row_sql_date

    init_database()
    db = SessionLocalFactory()
    try:
        rows = db.execute(
            text(
                """
                SELECT rebalance_date, COUNT(*) AS cnt
                FROM strategy_positions
                WHERE strategy_id=:sid
                GROUP BY rebalance_date
                ORDER BY rebalance_date
                """
            ),
            {"sid": SID},
        ).mappings().all()
        counts: dict[date, int] = {}
        total = 0
        min_rb = max_rb = None
        for r in rows:
            rb = _row_sql_date(r["rebalance_date"])
            if rb is None:
                continue
            c = int(r["cnt"] or 0)
            counts[rb] = c
            total += c
            min_rb = rb if min_rb is None else min(min_rb, rb)
            max_rb = rb if max_rb is None else max(max_rb, rb)
        return counts, total, min_rb, max_rb
    finally:
        db.close()


def cumulative_at_date(period_counts: dict[date, int], through: date) -> int:
    return sum(c for rb, c in period_counts.items() if rb <= through)


def main() -> None:
    excel_path = Path(sys.argv[1] if len(sys.argv) > 1 else EXCEL_DEFAULT)
    if not excel_path.is_file():
        print(f"文件不存在: {excel_path}")
        sys.exit(1)

    print(f"Excel: {excel_path}")
    ex_counts, ex_total = excel_period_counts(excel_path)
    ex_periods = len(ex_counts)
    ex_min = min(ex_counts)
    ex_max = max(ex_counts)

    db_counts, db_total, db_min, db_max = db_period_counts()

    print("\n=== 汇总 ===")
    print(f"Excel: {ex_total} 行, {ex_periods} 个调仓日, {ex_min} ~ {ex_max}")
    print(f"Turso: {db_total} 行, {len(db_counts)} 个调仓日, {db_min} ~ {db_max}")

    if db_max:
        ex_cum_at_db_max = cumulative_at_date(ex_counts, db_max)
        print(
            f"\n若库内最后日期 {db_max} 正确，累计行数应约 {ex_cum_at_db_max}；"
            f"实际库内 {db_total}，差 {ex_cum_at_db_max - db_total}"
        )

    mismatches: list[tuple[date, int, int]] = []
    all_dates = sorted(set(ex_counts) | set(db_counts))
    for rb in all_dates:
        ec = ex_counts.get(rb, 0)
        dc = db_counts.get(rb, 0)
        if ec != dc:
            mismatches.append((rb, ec, dc))

    print(f"\n=== 调仓日不一致（共 {len(mismatches)} 期）===")
    for rb, ec, dc in mismatches[:15]:
        print(f"  {rb}: Excel={ec} DB={dc} Δ={dc - ec}")
    if len(mismatches) > 15:
        print(f"  … 另有 {len(mismatches) - 15} 期")

    if mismatches:
        first_bad = mismatches[0][0]
        print(f"\n建议续传/重导起点（首期不一致）: {first_bad}")
        print(
            f"  该日 Excel 累计至当日: {cumulative_at_date(ex_counts, first_bad)} 行"
        )

    if db_max and db_max in ex_counts:
        print(f"\n库内最后一期 {db_max}: Excel={ex_counts[db_max]} DB={db_counts.get(db_max, 0)}")

    ex_only = sorted(set(ex_counts) - set(db_counts))
    db_only = sorted(set(db_counts) - set(ex_counts))
    if ex_only:
        print(f"\nExcel 有但库内无的调仓日（前5）: {ex_only[:5]} … 共 {len(ex_only)}")
    if db_only:
        print(f"库内有但 Excel 无的调仓日: {db_only[:5]}")


if __name__ == "__main__":
    main()

File: scripts/compare_cl1_excel_db.mjs
#!/usr/bin/env node
/**
 * 对照 CL1 Excel 与 Turso（列：调整日期、证券代码，与导入逻辑一致）。
 * 用法: node scripts/compare_cl1_excel_db.mjs [excel路径]
 */
import { createClient } from "@libsql/client";
import { readFile } from "node:fs/promises";
import XLSX from "xlsx";

const EXCEL =
  process.argv[2] ||
  "D:\\展示策略\\展示策略\\展示策略-沪深300指数增强.xlsx";
const SID = "CL1";
const MAX_COL = 5;

function parseDate(v) {
  if (v == null || v === "") return null;
  if (v instanceof Date && !isNaN(v)) return v.toISOString().slice(0, 10);
  if (typeof v === "number") {
    const epoch = new Date(Date.UTC(1899, 11, 30));
    const d = new Date(epoch.getTime() + v * 86400000);
    return d.toISOString().slice(0, 10);
  }
  const s = String(v).trim();
  if (!s) return null;
  const m = s.match(/^(\d{4})[-/](\d{1,2})[-/](\d{1,2})/);
  if (m) return `${m[1]}-${m[2].padStart(2, "0")}-${m[3].padStart(2, "0")}`;
  const d = new Date(s);
  if (!isNaN(d)) return d.toISOString().slice(0, 10);
  return null;
}

function normCode(raw) {
  if (raw == null || raw === "") return null;
  let t =
    typeof raw === "number"
      ? String(Number.isInteger(raw) ? raw : Math.trunc(raw))
      : String(raw).trim().toUpperCase();
  if (t.endsWith(".0") && /^[0-9]+\.0$/.test(t)) t = t.slice(0, -2);
  if (!t) return null;
  if (t.includes(".")) return t;
  if (t.length === 6 && t.startsWith("6")) return `${t}.SH`;
  if (t.length === 6) return `${t}.SZ`;
  return t;
}

async function excelPeriodCounts(path) {
  const buf = await readFile(path);
  const wb = XLSX.read(buf, { type: "buffer", cellDates: true });
  const ws = wb.Sheets[wb.SheetNames[0]];
  const rows = XLSX.utils.sheet_to_json(ws, { header: 1, defval: null });
  if (!rows.length) throw new Error("空表");
  const header = rows[0].slice(0, MAX_COL).map((c) => String(c ?? "").trim());
  const iDt = header.indexOf("调整日期");
  const iCode = header.indexOf("证券代码");
  if (iDt < 0 || iCode < 0) throw new Error(`缺少列 header=${header.join(",")}`);
  const counts = new Map();
  let total = 0;
  let lastRb = null;
  for (let r = 1; r < rows.length; r++) {
    const cells = rows[r].slice(0, MAX_COL);
    let rb = parseDate(cells[iDt]);
    const code = normCode(cells[iCode]);
    if (!code) continue;
    if (!rb) rb = lastRb;
    if (!rb) continue;
    lastRb = rb;
    counts.set(rb, (counts.get(rb) || 0) + 1);
    total += 1;
  }
  return { counts, total };
}

async function dbPeriodCounts() {
  const url = process.env.TURSO_DATABASE_URL;
  const authToken = process.env.TURSO_AUTH_TOKEN;
  if (!url || !authToken) throw new Error("需要 TURSO_DATABASE_URL / TURSO_AUTH_TOKEN");
  const db = createClient({ url, authToken });
  const rows = await db.execute({
    sql: `SELECT rebalance_date AS rb, COUNT(*) AS cnt
          FROM strategy_positions WHERE strategy_id=? GROUP BY rebalance_date ORDER BY rb`,
    args: [SID],
  });
  const counts = new Map();
  let total = 0;
  let minRb = null;
  let maxRb = null;
  for (const r of rows.rows) {
    const rb = String(r.rb).slice(0, 10);
    const c = Number(r.cnt);
    counts.set(rb, c);
    total += c;
    if (!minRb || rb < minRb) minRb = rb;
    if (!maxRb || rb > maxRb) maxRb = rb;
  }
  return { counts, total, minRb, maxRb };
}

function cumThrough(counts, through) {
  let n = 0;
  for (const [rb, c] of counts) if (rb <= through) n += c;
  return n;
}

const ex = await excelPeriodCounts(EXCEL);
const db = await dbPeriodCounts();
const exDates = [...ex.counts.keys()].sort();
const dbDates = [...db.counts.keys()].sort();
const exMin = exDates[0];
const exMax = exDates[exDates.length - 1];

console.log("Excel:", EXCEL);
console.log("\n=== 汇总 ===");
console.log(
  `Excel: ${ex.total} 行, ${ex.counts.size} 期, ${exMin} ~ ${exMax}`
);
console.log(
  `Turso: ${db.total} 行, ${db.counts.size} 期, ${db.minRb} ~ ${db.maxRb}`
);

if (db.maxRb) {
  const expect = cumThrough(ex.counts, db.maxRb);
  console.log(
    `\n库内最后日 ${db.maxRb}：按 Excel 累计应约 ${expect} 行，实际 ${db.total}，差 ${expect - db.total}`
  );
}

const all = [...new Set([...ex.counts.keys(), ...db.counts.keys()])].sort();
const bad = [];
for (const rb of all) {
  const ec = ex.counts.get(rb) || 0;
  const dc = db.counts.get(rb) || 0;
  if (ec !== dc) bad.push({ rb, ec, dc });
}
console.log(`\n=== 调仓日不一致（${bad.length} 期）===`);
bad.slice(0, 20).forEach(({ rb, ec, dc }) =>
  console.log(`  ${rb}: Excel=${ec} DB=${dc} Δ=${dc - ec}`)
);
if (bad.length > 20) console.log(`  … 另有 ${bad.length - 20} 期`);

if (bad.length) {
  const first = bad[0].rb;
  console.log(`\n建议重导起点（首期不一致）: ${first}`);
  console.log(`  至该日 Excel 累计: ${cumThrough(ex.counts, first)} 行`);
}

if (db.maxRb && ex.counts.has(db.maxRb)) {
  console.log(
    `\n库内最后一期 ${db.maxRb}: Excel=${ex.counts.get(db.maxRb)} DB=${db.counts.get(db.maxRb) || 0}`
  );
}
